File: PCCP - Game Of Bricks/src/gamehandler.cc
#include "header.h"
#include "gamehandler.h"
#include <string.h>

GameHandler :: GameHandler()
{
	head = NULL;
	breakableBrks = 0;
	gmspd = GMSPD_SLOW;
}
void GameHandler :: clearLoadedLevel()
{
	Brick *curr;
	while (head != NULL)
	{
		curr = head;
		head = head->get_next();
		delete (curr);
	}
	breakableBrks = 0;
}
void GameHandler :: loadLevel (char *levelName)
{
	FILE *fp;
	int N;
	int brk_x, brk_y, brk_color, brk_hits;
	Brick *tail, *curr;
	clearLoadedLevel();
	tail = NULL;		
	fp = fopen(levelName, "r");
	fscanf(fp,"%d ",&N);
	for (int i=0; i<N; i++)
	{
		fscanf(fp, "%d ", &brk_x);
		fscanf(fp, "%d ", &brk_y);			
		fscanf(fp, "%d ", &brk_color);				
		fscanf(fp, "%d\n",&brk_hits);
		curr = new Brick(brk_x, brk_y, brk_color, brk_hits);
		if (!curr->isRigid())
			breakableBrks++;
		if (head==NULL)
		{
			head = curr;
			tail = curr;
		}
		else
		{
			tail->set_next(curr);
			tail = curr;
		}
	}
} // loadFile

void GameHandler :: loadBricks()
{
	Brick *curr;
	for (curr=head;curr!=NULL;curr=curr->get_next())
		curr->drawBrick();
} // loadBricks

void GameHandler :: removeBrick(Brick *prev, Brick *curr)
{
	if (curr==prev)
		head = head->get_next();
	else
		prev->set_next(curr->get_next());
	delete(curr);
} // removeBrick

int GameHandler :: askOptions ()
{

	int option;
	clear();
	echo();

	player.set_score(0);
	breakableBrks = 0;

	mvprintw(0, 0, "0 : Game Settings\n");
	printw  ("1 : Use Default Settings\n");
	printw  ("2 : Quit\n");
	printw  ("Please Enter your Option here : ");
	refresh();
	do
	{
		scanw ("%d", &option);
		switch (option)
		{
			case 0	: doGameSettings();	break;
			case 1	: doDftSettings();	break;
			case 2	: return QUIT;
			default	: printw  ("Please Enter Correct Option (0/1/2) : ");	refresh();
		}
	} while (option != 0 && option != 1 && option != 2);
	return OK;
} // askOptions

void GameHandler :: doDftSettings()
{
	// Difficulty Level - Default is EASY
	player.set_lives (DFT_LIVES);		// DFT_LIVES = 5
	
	// Paddle Speed - Default is SLOW
	pad.set_speed (PAD_SLOW);
	
	// Paddle Length - Default is LONG
	pad.set_length (PAD_LONG);
	
	// Game Speed - Default is SLOW
	gmspd = GMSPD_SLOW;
} // doDftSettings

void GameHandler :: doGameSettings()
{
	player.readName();
	echo();
	// Difficulty Level
	int option;
	clear();
	printw ("Difficulty Level\n");
	printw ("0 : EASY\n");
	printw ("1 : HARD\n");
	printw ("Please Enter your Option here : ");
	refresh();
	do
	{
		scanw ("%d", &option);
		switch (option)
		{
			case EASY : 	pad.set_length (PAD_LONG);
					player.set_lives (MORE);
					break;
			case HARD :	pad.set_length (PAD_SHORT);
					player.set_lives (LESS);
					break;
			default	: 	printw ("Please Enter Correct Option (0/1) : ");	refresh();
		}
	} while (option != EASY && option != HARD);
	
	// Paddle Speed
	clear();
	printw ("Paddle Speed\n");
	printw ("0 : SLOW\n");
	printw ("1 : FAST\n");
	printw ("Please Enter your Option here : ");
	refresh();
	do
	{
		scanw ("%d", &option);
		switch (option)
		{
			case SLOW : 	pad.set_speed(PAD_SLOW);
					break;
			case FAST :	pad.set_speed(PAD_FAST);
					break;
			default	: 	printw ("Please Enter Correct Option (0/1) : ");	refresh();
		}
	} while (option != SLOW && option != FAST);


	// Game Speed
	clear();
	printw ("Game Speed\n");
	printw ("0 : SLOW\n");
	printw ("1 : FAST\n");
	printw ("Please Enter your Option here : ");
	refresh();
	do
	{
		scanw ("%d", &option);
		switch (option)
		{
			case SLOW : 	gmspd = GMSPD_SLOW;
					break;
			case FAST :	gmspd = GMSPD_FAST;
					break;
			default	: 	printw ("Please Enter Correct Option (0/1) : ");	refresh();
		}
	} while (option != SLOW && option != FAST);

} // doGameSettings

int GameHandler :: play (char *levelName)
{
	loadLevel(levelName);
	int key = initialise();
	if (key == QUIT || key == PREV_LEVEL || key == SKIP_LEVEL)
		return key;
	while (!isLevelComplete())
	{
		key = chkKeyPressed();
		if (key == QUIT || key == PREV_LEVEL || key == SKIP_LEVEL)
			return key;
		ball.moveBall();
		player.displayInfo();
		//mvprintw (MSG_POSY,MSG_POSX,"Breakable = %d\t",breakableBrks);
		//refresh();
		int status = check();
		if (status == NOT_OK)
		{
			// Player losses one turn
			if (player.hasLife())
			{
				player.reduceLives();
				mvprintw (MSG_POSY, MSG_POSX, "Sorry, You lost one chance\n");
				printw ("Press any key to start\n");
				refresh();
				timeout (HALT);
				getch();
				key = initialise();
				if (key == QUIT) return QUIT;
			}
			else
			{
				// Game Over
				gameOver();
				return QUIT;
			}
		}
		delay();
	}
	return (strcmp(levelName, LAST_LEVEL)==0) ? GAME_COMPLETED : LEVEL_COMPLETED;
} // play
int GameHandler :: initialise()
{
	clear();
	int key = 0;
	noecho();
	player.displayInfo();
	pad.init();
	ball.init();
	ball.showDir();
	loadBricks();
	do
	{
		key = wgetch(stdscr);
		if (key == KEY_LEFT)
		{
			pad.movePaddle (LEFT);
			ball.moveBall (pad.get_startx(), pad.get_length());
		}
		if (key == KEY_RIGHT)
		{
			pad.movePaddle (RIGHT);
			ball.moveBall (pad.get_startx(), pad.get_length());
		}
		if (key == KEY_UP)
			ball.changeDir();
		if (key == QUIT)
		{
			mvprintw (MSG_POSY, MSG_POSX, "Do you want to quit (y/n)? : ");
			refresh();
			timeout (HALT);
			key = getch();
			if (key == YES)
				return QUIT;
			else
			{
				move(MSG_POSY, MSG_POSX);
				clrtoeol ();
			}
			refresh();
		}

	} while (key != SPACE_BAR && key != PREV_LEVEL && key != SKIP_LEVEL);
	return key;
} // initialise


int GameHandler :: isLevelComplete()
{
	return (breakableBrks == 0);
}

void GameHandler :: delay()
{
	for (double i=0.0; i<DELAY; i+=gmspd);
}

void GameHandler :: gameOver()
{
	clear();
	mvprintw (0, 0, "GAME OVER!\n");
	printw ("Your Score is : %d\n", player.get_score());
	printw ("Press any key to continue\n");
	refresh();
	timeout (HALT);
	getch();
} // Game Over


int GameHandler :: check()
{
	int hor_ed = ball.isHitHorScrEd();
	int ver_ed = ball.isHitVerScrEd();

	// Ball Hits the Screen Corner
	if (hor_ed && ver_ed)
	{
		ball.revBallDir();
		return OK;
	}
	
	// Ball Hits only the Horizontal Screen Edge (i.e. Top Edge)
	if (hor_ed)
	{
		ball.flipyi();
		return OK;
	}

	// If ball is about to hit the Paddle
	int near_pad = ball.isNearPad();

	// Ball hits only the Vertical Screen Edge
	if (ver_ed && !near_pad)
	{
		ball.flipxi();
		return OK;
	}

	// Ball hits the pad corner or the pad top
	int pad_cor = pad.isBallHitPadCor(ball);
	int pad_top = pad.isBallHitPadTop(ball);
	if (near_pad)
	{
	/*	if (!ver_ed)
			ball.flipyi();
		else	ball.revBallDir();
		return OK;
	*/
		// Ball hits only the top of the paddle
		if (pad_top && !ver_ed)
		{
			ball.flipyi();
			beep();
			return OK;
		}
		
		// Ball hits the only pad_cor or pad_top and ver_ed
		// or pad_cor and ver_ed
		if ((pad_cor && !ver_ed) || (pad_top && ver_ed))
		{
			ball.revBallDir();
			beep();
			return OK;
		}
		return NOT_OK;
	}
	else
		chkHitsToBrks();
	return OK;
} // check

void GameHandler :: chkHitsToBrks()
{
	Brick *prev, *curr;
	curr = head;
	prev = head;
	
	int flag = FALSE;	// To check whether Ball has hit a Brick
	for (; curr!=NULL; curr=curr->get_next())
	{
		if (curr->isBallHitBrkCor(ball))
		{
			ball.revBallDir();
			flag = TRUE;
		}
		if (curr->isBallHitBrkHorWall(ball))
		{
			ball.flipyi();
			flag = TRUE;
		}
		if (curr->isBallHitBrkVerWall(ball))
		{
			ball.flipxi();
			flag = TRUE;
		}
		if (flag)
		{
			int score = 0;
			if (!curr->isRigid())
			{
				if (!curr->isAboutToBreak())
				{
					score = 60;
					curr->changeColor();
					curr->reduceHits();
				}
				else
				{
					score = 10;
					breakableBrks--;
					curr->eraseBrick();
					removeBrick(prev, curr);
				}
				player.updateScore(score);
			}
			//beep();
			break;
		}
		prev = curr;
	} // for
	
} // chkHitsToBrks


int GameHandler :: chkKeyPressed()
{
	timeout(0);
	int key = wgetch(stdscr);
	if (key == KEY_LEFT)
		pad.movePaddle(LEFT);
	if (key == KEY_RIGHT)
		pad.movePaddle(RIGHT);
	if (key == PAUSE)
	{
		timeout(HALT);
		getch();
		timeout(0);
	}
	if (key == SPACE_BAR)
	{
		if (!ball.isHitHorScrEd())
			if (ball.isDirLeftDown() || ball.isDirRightDown())
				ball.flipyi();
	}
	if (key == QUIT)
	{
		mvprintw (MSG_POSY, 0, "Quit (y or n)? : ");
		refresh();
		timeout (HALT);
		int option = getch();
		if (option == YES) 
			return QUIT;
		else
		{
			move (MSG_POSY, MSG_POSX);
			clrtoeol ();
			refresh();
			return option;
		}
	}
	return key;
}
void GameHandler :: gameCompleted ()
{
	clear();
	attron (A_BLINK);
	mvprintw (0,0,"CONGRAGULATIONS !!\n");
	printw ("You have successfully completed the GAME OF BRICKS");
	refresh();
	getch();
}

int  main()
{ 	
	char filename[20];
	char *levels[] = { 	"level01.txt", "level02.txt", "level03.txt",
				"level04.txt", "level05.txt", "level06.txt",
				"level07.txt", "level08.txt", "level09.txt",
				"level10.txt"
			};
	/*			
	printf("Enter the file name:");
	scanf("%s",&filename);
	*/
	initscr();
   	noecho();
	cbreak();

	keypad(stdscr,TRUE);
	attron(A_BOLD);
	start_color();
	init_pair(BRK_HITS1, COLOR_RED,    COLOR_RED);	  // BRICK HITS = 1
	init_pair(BRK_HITS2, COLOR_YELLOW, COLOR_YELLOW); // BRICK HITS = 2
	init_pair(BRK_RIGID, COLOR_CYAN,   COLOR_CYAN);	  // Rigid Brick
	init_pair(PAD_COLOR, COLOR_BLUE,   COLOR_BLUE);	  // Paddle
	init_pair(BALL_COLOR,COLOR_YELLOW,  COLOR_BLACK);  // Ball

	GameHandler gmhdr;
	int option;
	do
	{
		option = gmhdr.askOptions();
		if (option == QUIT) break;
		int status;
		for (int i=0; i<10; i++)
		{	
			status = gmhdr.play(*(levels+i));
			if (status == QUIT)
				break;
			if (status == PREV_LEVEL)
				if (i != 0)
					i-=2;
				else i = -1;
			if (status == SKIP_LEVEL || status == LEVEL_COMPLETED)
				if(strcmp(*(levels+i), LAST_LEVEL) != 0)
					continue;
				else
					status = GAME_COMPLETED;
			clear();
			refresh();
		}
		if (status == GAME_COMPLETED)
		{
			gmhdr.gameCompleted();
			break;
		}
	} while (option != QUIT);
/*
	gmhdr.play(filename);
	clear();
*/
	gmhdr.clearLoadedLevel();
//	getch();
	endwin();
	return 1;
} // main

File: OOAD - Game Of Bricks/GameHandler.java
/************************************************************************************/
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import ncst.pgdst.*;
import java.io.File;
/************************************************************************************/
public class GameHandler extends JApplet implements KeyListener, ActionListener, CONSTANTS
{
    private final static Color bg = Color.black;
    private static Rectangle gameBorder;
	private JFrame frame;
	private Dimension dim;
	Timer timer;

	// for double buffering
	private Image dbImage;
	private Graphics dbg;

	private boolean initialScreen, gameInitialised, gameStarted, gamePaused;

	/* whether to clear the screen or not
		screen is cleared 
			- when initialising game from Initial Screen 
			- loading next or previous level
			- showing Initial Screen if user quits the running game in between
	*/
	private boolean clearScreen;

	private Player player;
	private Ball ball;
	private Paddle pad;
	private Setting set;
	private Brick brkHead;
	private String currLevelFile;
	private int currLevelNo;
	private int no_of_bricks;
	private int breakableBricks;

/*==========================================================================*/
	public GameHandler() throws IOException
	{
		// create a frame to hold the game graphics
		frame = new JFrame("Game of Bricks");
		frame.addWindowListener(new WindowAdapter() {
			public void windowClosing(WindowEvent e) {System.exit(0);}
		});
		frame.getContentPane().add("Center", this);
		init();
		frame.pack();
		frame.setResizable(false);
		dim = frame.getToolkit().getScreenSize();
		frame.setSize(dim);
		gameBorder = new Rectangle(10, 110, dim.width-30, dim.height-150);
		
		// to show the inital screen with the options
		initialScreen = true;
		clearScreen = gameInitialised = gameStarted = gamePaused = false;

		// initalise other components of the game
		player = new Player(dim, "Player");
		set = new Setting();
		ball = new Ball();
		pad = new Paddle();
		brkHead = null;
		currLevelNo = 0;
		currLevelFile = LEVEL_FILE[currLevelNo];
		timer = new Timer (GAMESPEED_MEDIUM, this);
		timer.start();
		
        frame.show();			// this calls the paint
		addKeyListener(this);

		// ask the player his name
		String name = (String)JOptionPane.showInputDialog(GameHandler.this,
			"Please enter your name : ",
			BOX_TITLE, JOptionPane.INFORMATION_MESSAGE, null, null,"Player");
		player.setName(name);
	}
/*==========================================================================*/
	// this method will load level data from the specifed file and create a linked list of bricks
	// present in that level
	public void loadLevel(String filename) throws IOException
	{
		// remove any old bricks of previous level present in the linked list 
		brkHead = null;
		breakableBricks = 0;

		// open the current level file
		File file = new File(filename);
		SimpleInput sin = new SimpleInput(file);

		// read data from file and create linked list of bricks
		no_of_bricks = sin.readInt();
		for (int i=1; i<=no_of_bricks; i++)
		{
			int x = sin.readInt();
			int y = sin.readInt();
			int color = sin.readInt();
			int hits = sin.readInt();
			// count how many bricks are breakable
			if (hits != -1)
			{
				breakableBricks++;
			}
			// add brick to the linked list
			if (brkHead == null)
			{
				brkHead = new Brick(x, y, color, hits);
			}
			else
			{
				addBrickToList(new Brick(x, y, color, hits));
			}
		}
	} //loadLevel
/*==========================================================================*/
	// add a brick at the end of the linked list of bricks
	public void addBrickToList(Brick newBrick)
	{
		Brick curr = brkHead;
		while (curr.next != null)
		{
			curr = curr.next;
		}
		curr.next = newBrick;
	}
/*==========================================================================*/	
	// display all the bricks present in the linked list
	public void displayBricks(Graphics g)
	{
		Brick curr = brkHead;
		while (curr != null)
		{
			curr.drawBrick(g);
			curr = curr.next;
		}
	}
/*==========================================================================*/
	public void init() 
	{
		// Initialize background color
		setBackground(bg);
	}
/*==========================================================================*/	
	// it shows the inital screen with some options like 1)Start game, 2) Change setting, etc.
	public void showInitialScreen (Graphics g)
	{
		// erase previous screen
		int maxWidth = (int)dim.getWidth();
		int maxHeight = (int)dim.getHeight();
		g.setColor(Color.black);
		g.fillRect(0, 0, maxWidth, maxHeight);

		Font fancyFont = new Font("Serif", Font.BOLD | Font.ITALIC, 50);
		g.setColor(Color.red);
		g.drawRect(START_X,START_Y,maxWidth-OFF_X,maxHeight-OFF_Y);
		String Title="Game of Bricks";
		String key1="Press 'S' or 's' to start the game";
		String key2="Press 'C' or 'c' to change settings";
		String key3="Press H or 'h'to see Help for the Game";
		String key4="Press 'Q' or 'q' to quit game";
		Font f1=new Font("Serif",Font.PLAIN | Font.ITALIC,25);
		g.setFont(fancyFont);
		g.setColor(Color.green);
		g.drawString(Title,START_X+maxWidth/4-25,START_Y+maxHeight/4-25);
		g.setFont(f1);
		g.setColor(Color.blue);
		g.drawString(key1,START_X+maxWidth/4-OFFSET,START_Y+maxHeight/4+OFFSET);
		g.drawString(key2,START_X+maxWidth/4-OFFSET,START_Y+maxHeight/4+2*OFFSET);
		g.drawString(key3,START_X+maxWidth/4-OFFSET,START_Y+maxHeight/4+3*OFFSET);
		g.drawString(key4,START_X+maxWidth/4-OFFSET,START_Y+maxHeight/4+4*OFFSET);
	} // showInitialScreen
/*==========================================================================*/
    public void paint(Graphics g) 
	{
		if (initialScreen)
		{
			showInitialScreen(g);
		}
		else if (gameInitialised)
		{
			if (clearScreen)
			{
				clearScreen = false;
				g.setColor(Color.black);
				g.fillRect(0, 0, dim.width, dim.height);
			}
			player.displayPlayerInfo(g, currLevelNo+1);
			g.setColor(Color.white);
			g.drawRect(gameBorder.x, gameBorder.y, gameBorder.width, gameBorder.height);
			displayBricks(g);

			pad.drawPaddle(g);
			ball.drawBall(g);
			ball.showDir(g);
		}
		else if (gameStarted)
		{
			player.displayPlayerInfo(g, currLevelNo+1);
			g.setColor(Color.white);
			g.drawRect(gameBorder.x, gameBorder.y, gameBorder.width, gameBorder.height);
			displayBricks(g);

			pad.drawPaddle(g);
			ball.drawBall(g);
		}
	} // paint
/*==========================================================================*/
	/** Update - Method, implements double buffering */
	public void update (Graphics g)
	{ 
		// initialize buffer
		if (dbImage == null)
		{ 
			dbImage = createImage (this.getSize().width, this.getSize().height);
			dbg = dbImage.getGraphics ();
		}
		// clear screen in background
		dbg.setColor (getBackground ());
		dbg.fillRect (0, 0, this.getSize().width, this.getSize().height);

		// draw elements in background
		dbg.setColor (getForeground());
		paint(dbg);

		// draw image on the screen
		g.drawImage (dbImage, 0, 0, this);
	} // update
/*==========================================================================*/
	public void actionPerformed (ActionEvent ae)
	{
		if (gameStarted)
		{
			if (!isLevelCompleted())
			{
				ball.moveBall();
				int status = detectCollision();
				if (status == NOT_OK) // player has lost one turn
				{
					if (player.hasLife())
					{
						// - stop the timer
						timer.stop();
						player.reduceLives();
						// show a message box that player have lost once life
						JOptionPane.showMessageDialog(GameHandler.this, "You lost one life.",
							BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
						gameInitialised = true;
						gameStarted = false;
						initGame();
						timer.setDelay(set.getGameSpeed());
						timer.restart();
					}
					else		// player has lost all the lives and hence game is over
					{
						gameOver();
					}
				}
			}
			else // level completed and hence load next level
			{
				// if level completed is the last level
				if (currLevelNo == LAST_LEVEL)
				{
					gameCompleted();
				}
				else
				{
					// show message that level is completed
					JOptionPane.showMessageDialog(GameHandler.this, "Level Completed.",
							BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
					interpretKey(NEXT_LEVEL);
				}
			}
		} // if (gameStarted)
		repaint();
	} // actionPerformed
/*==========================================================================*/
	public int detectCollision()
	{
		// detect collision of the ball with the screen edges
		boolean hor_ed = ball.isHitHorScrEd();
		boolean ver_ed = ball.isHitVerScrEd();
		// ball hits screen corner - TOP_LEFT or TOP_RIGHT corner
		if (hor_ed && ver_ed)
		{
			ball.revBallDir();
			return OK;
		}
		// ball hits only TOP_EDGE
		if (hor_ed)
		{
			ball.flipyi();
			return OK;
		}
		
		boolean near_pad = ball.isNearPaddle();
		// ball hits only LEFT_EDGE or RIGHT_EDGE
		if (ver_ed && !near_pad)
		{
			ball.flipxi();
			return OK;
		}

		// ball hits the the paddle top edge
		boolean pad_top = pad.isBallHitPadTop	(ball);
		if (near_pad)
		{
			// ball hits only the paddle top edge
			if (pad_top && !ver_ed)
			{
				ball.flipyi();
				return OK;
			}
			return NOT_OK;
		}
		else
		{
			checkHitsToBrick();
		}
		return OK;
	} // detectCollision
/*==========================================================================*/
	// traverse through the linked list of bricks and check whether ball hits a brick
	public void checkHitsToBrick()
	{
		Brick curr = brkHead;
		Brick prev = brkHead;
		boolean flag = false;
		while (curr != null)
		{
			if (curr.isBallHitBrkHorWall(ball))
			{
				ball.flipyi();
				flag = true;
			}
			if (curr.isBallHitBrkVerWall(ball))
			{
				ball.flipxi();
				flag = true;
			}
			if (flag)
			{
				int score = 0;
				if (!curr.isRigid())
				{
					score = curr.getPoints();
					if (!curr.isAboutToBreak())	// change the color of the bricks
					{
						curr.reduceHits();
						curr.setColor(curr.getHits());
					}
					else	// brick is to be destroyed
					{
						breakableBricks--;
						curr.setColor(0);
						removeBrick(prev, curr);
					}
					player.updateScore(score);
				}
				break;
			}
			prev = curr;
			curr = curr.next;
		}
	} // checkHitsToBrick
/*==========================================================================*/
	public void removeBrick(Brick prev, Brick curr)
	{
		if (prev == curr)	// brick to be removed is the first brick
		{
			brkHead = brkHead.next;
		}
		else
		{
			prev.next = curr.next;
		}
	} // removeBrick
/*==========================================================================*/
	public void gameOver()
	{
		/*	 - stop the timer - timer.stop(); 
			 - show a message box with the player's score 
			 - then show initial screen
			 - initialScreen = clearScreen = true;
			 - gameInitialised =  gameStarted = false;
			 - initGame();
		*/
		timer.stop();
		JOptionPane pane = new JOptionPane("Game Over");
		JOptionPane.showMessageDialog(GameHandler.this, "Game Over! Your Score is "+player.getScore(),
			BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
		initialScreen = clearScreen = true;
		gameInitialised =  gameStarted = false;
		initGame();
		player.setScore(0);
		player.setLives(MAX_LIVES);
	}
/*==========================================================================*/
	public void gameCompleted()
	{
		/* - stop the timer - timer.stop();
			- show a message box with the player's score and a congrats message
			- then show initial screen
			- initialScreen = clearScreen = true;
			- gameInitialised =  gameStarted = false;
			- initGame();
			- set player's info
			- set level no = FIRST_LEVEL
		*/
		timer.stop();
		JOptionPane pane = new JOptionPane("Game Completed");
		JOptionPane.showMessageDialog(GameHandler.this, "Congrats!! You have completed the Game of Bricks!!! Your Score is "+player.getScore(),
			BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
		initialScreen = true;
		clearScreen = true;
		gameInitialised =  gameStarted = false;
		initGame();
		player.setLives(MAX_LIVES);
		player.setScore(0);
		currLevelNo = FIRST_LEVEL;
	}
/*==========================================================================*/
	public boolean isLevelCompleted()
	{
		return (breakableBricks == 0);
	}
/*==========================================================================*/
	// Invoked when a key has been pressed. 
	public void keyPressed(KeyEvent e) 
	{
		int key = e.getKeyCode();
		interpretKey (key);
	}
/*==========================================================================*/
	// total no. of keys to be interpreted are = 11
	public void interpretKey (int key) 
	{
		switch (key)
		{
			case START_GAME				:	if (initialScreen)
																{
																	clearScreen = true;
																	initialScreen = false;
																	gameInitialised = true;
																	try { loadLevel(currLevelFile); }
																	catch (Exception ex) {}
																	repaint();
																}
																return;
			case RUN_GAME					:	if (gameInitialised)
																{
																	gameStarted = true;
																	clearScreen = false;
																	initialScreen = false;
																	gameInitialised = false;
																	timer.setDelay(set.getGameSpeed());
																	timer.restart();
																}
																return;

			// always interpret this but respond differently depending upon the game status
			case CHANGE_SETTING	:	if (initialScreen || gameInitialised)
																{
																	set.changeSetting();
																}
																else if (gameStarted)
																{
																	timer.stop();
																	set.changeSetting();
																	timer.setDelay(set.getGameSpeed());
																	timer.restart();
																}
																return;

			// always interpret this but respond differently depending upon the game status
			case HELP								:	if (initialScreen || gameInitialised)
																{
																	JOptionPane pane = new JOptionPane(HELP_MSG);
																	JOptionPane.showMessageDialog(GameHandler.this, HELP_MSG,
																		BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
																}
																else if (gameStarted)
																{
																	timer.stop();
																	JOptionPane pane = new JOptionPane(HELP_MSG);
																	JOptionPane.showMessageDialog(GameHandler.this, HELP_MSG,
																		BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
																	timer.setDelay(set.getGameSpeed());
																	timer.restart();
																}
																return;

			case LEFT_ARROW				:	if (gameInitialised || gameStarted)
																{
																	pad.movePaddle(LEFT, set.getPadSpeed());
																	if (gameInitialised)
																	{
																		ball.moveBall(pad.getStartx(), pad.getLength());
																	}
																	repaint();
																}
																return;
			case RIGHT_ARROW			:	if (gameInitialised || gameStarted)
																{
																	pad.movePaddle(RIGHT, set.getPadSpeed());
																	if (gameInitialised)
																	{
																		ball.moveBall(pad.getStartx(), pad.getLength());
																	}
																	repaint();
																}
																return;
			case UP_ARROW					:	if (gameInitialised)
																{
																	ball.changeDir();
																	repaint();
																}
																return;
			case PAUSE							:	if (gameStarted)
																{
																	timer.stop();
																	JOptionPane pane = new JOptionPane("Game Paused");
																	JOptionPane.showMessageDialog(GameHandler.this, "Game Paused",
																		BOX_TITLE, JOptionPane.INFORMATION_MESSAGE);
																	timer.start();
																}
																return;
			case NEXT_LEVEL				:	if (!initialScreen)
																{
																	if (currLevelNo != LAST_LEVEL)
																	{
																		// initGame() - re-init ball, paddle co-ordinates
																		clearScreen = true;
																		gameStarted = false;
																		gameInitialised = true;
																		initGame();
																		currLevelNo++;
																		try { loadLevel(LEVEL_FILE[currLevelNo]); }
																		catch (Exception ex) {}
																		if (timer.isRunning())
																		{
																			timer.stop();
																		}
																		repaint();
																	}
																}
																return;
			case PREV_LEVEL				:	if (!initialScreen)
																{
																	if (currLevelNo != FIRST_LEVEL)
																	{
																		// initGame() - re-init ball, paddle co-ordinates
																		clearScreen = true;
																		gameStarted = false;
																		gameInitialised = true;
																		initGame();
																		currLevelNo--;
																		try { loadLevel(LEVEL_FILE[currLevelNo]); }
																		catch (Exception ex) {}
																		if (timer.isRunning())
																		{
																			timer.stop();
																		}
																		repaint();
																	}
																}
																return;
			case QUIT								:	if (initialScreen)
																{
																	if (JOptionPane.showConfirmDialog(GameHandler.this,
																		"Do you want to quit this application?",
																		BOX_TITLE, JOptionPane.YES_NO_OPTION)
																		== JOptionPane.YES_OPTION)
																			System.exit(0);
																	// return;
																}
																else if (gameInitialised || gameStarted) 
																{
																	if (gameStarted)
																	{
																		timer.stop();
																	}
																	if (JOptionPane.showConfirmDialog(GameHandler.this,
																		"Do you want to quit this application?",
																		BOX_TITLE, JOptionPane.YES_NO_OPTION)
																		== JOptionPane.YES_OPTION)
																	{
																		// initGame() - re-init ball, paddle co-ordinates
																		initGame();
																		// init player's score to zero
																		player.setScore(0);
																		// init player's lives to MAX_LIVES
																		player.setLives(MAX_LIVES);
																		currLevelNo = 0;
																		gameStarted = false;
																		gameInitialised = false;
																		initialScreen = true;
																		clearScreen = true;
																		repaint();
																		// return;
																	}
																	else
																	{
																		if (gameStarted)
																		{
																			timer.setDelay(set.getGameSpeed());
																			timer.restart();
																		}
																		// return;
																	}
																}
																return;
			case CHEAT_CODE				:	if (gameStarted)
																{
																	if (ball.isDirLeftDown() || ball.isDirRightDown())
																	{
																		ball.flipyi();
																	}
																}
		} // switch
	} // interpretKey
/*==========================================================================*/	
	// it re-sets the ball and paddle co-ordinates to default values
	public void initGame()
	{
		ball.init();
		pad.init();
	}
/*==========================================================================*/
	// Invoked when a key has been released. 
	public void keyReleased(KeyEvent e) {}
/*==========================================================================*/
	// Invoked when a key has been typed. 
	public void keyTyped(KeyEvent e) {}
/*==========================================================================*/
	public static void main(String s[]) throws IOException
	{
        JApplet applet = new GameHandler();
    }
} // GameHandler
/************************************************************************************/
